Fix getColumnMeta() for GH-15287 Pdo\Pgsql::setAttribute(PDO::ATTR_PREFETCH, 0)

outtersg opened the pull request:
As stated in the UPGRADING, using the passthrough ("single-row") mode of libpq (introduced in #15287) forbids passing a new query while the current one's results have not been entirely consumed.
… But I didn't notice that ext/pdo_pgsql _internally_ used new queries to fetch metadata (example use case: a call to `getColumnMeta()` while `fetch()`ing row by row will interleave the `getColumnMeta()`-triggered internal query to the database, with the results fetching for the user-called query).

This PR **makes those internal calls return NULL for non-essential metadata**, instead of letting libpq abort the user-called query.

It moreover includes a small tweak to table oid-to-name translation, with a 1-slot cache.
This may by chance allow the internal call to return something instead of NULL,
but it will nonetheless avoid 30 server calls to get the table name of 30 columns of the same table.

devnexen commented on ext/pdo_pgsql/pgsql_statement.c:
```diff
@@ -702,12 +702,28 @@ static int pgsql_stmt_get_col(pdo_stmt_t *stmt, int colno, zval *result, enum pd
+	if (H->running_stmt) {
+		/* in single-row mode, libpq forbids passing a new query
+		 * while we're still flushing the current one's */
+		return 0;
```
nit: `return NULL` would be preferable (yes I know it s like that for the rest of the call).

outtersg replied:
I hesitated, so your advice helped tip the balance.

But by the way I had to recommit, as the CI detected my buggy confusion between "lazy statements" and "all statements": not selective enough I truncated the meta for the "classic" queries too.

… And so I thought of a third problem, cancelling the preceding query if _a new one_ entered the lazy fetching, instead of if _the preceding one_ was lazy and still unfinished.

outtersg commented:
I noticed this week-end that pgsql_driver.c too was full of PQexec (for internal needs).
So those calls too need to be protected the same way (by ensuring that no running statement has been let in a running state).

So [44afde3](https://github.com/php/php-src/pull/16249/commits/44afde3cb5a13cebbcacc13b38f5b825ff296bf7) exposes `pgsql_stmt_finish()`, and makes `pgsql_driver.c` use it extensively (only when the last run statement is in single-row / unbuffered / lazy fetch mode).

… But then I noticed some patterns that could be factorized _in_ `pgsql_stmt_finish()` (`while((pgsql_result = PQgetResult(H->server))) PQclear(pgsql_result);`).
I'm a bit stressed that factorized code could break things that presently work, so for now I just put a comment showing the mutualization opportunity (and the result is quite stable, I'm currently testing it with production cases),
but **should I fully explore finishing the code mutualization?**

devnexen commented:
> 
> … But then I noticed some patterns that could be factorized _in_ `pgsql_stmt_finish()` (`while((pgsql_result = PQgetResult(H->server))) PQclear(pgsql_result);`). I'm a bit stressed that factorized code could break things that presently work, so for now I just put a comment showing the mutualization opportunity (and the result is quite stable, I'm currently testing it with production cases), but **should I fully explore finishing the code mutualization?**

Sure, you can always explore in your own fork for a while and see how it goes. 8.5 release is in 1 year + :)

outtersg commented:
@devnexen wrote:

> Sure, you can always explore in your own fork for a while and see how it goes. 8.5 release is in 1 year + :)

That's a good point!

Now as I both rebased and amended, the changes are not that clear in the last diff :-\\
Here starts the [change to pdo_pgsql](https://github.com/php/php-src/compare/44afde3cb5a13cebbcacc13b38f5b825ff296bf7..a10ef194cb60acd2281e2be6c8a3bbd8f9713b00#diff-1921332920e2ebbd174954a81f2781b82adabff9cccae204a2e0dccf3c1ed62d).

outtersg commented:
Now that 8.4 is out, it's time to think of 8.5.

Although the nature of the PR (fixing an incompatibility between prefetch mode and metadata reading) could have made it a candidate for a 8.4.x,
the changes make it sufficently risky that I prefer keeping the 8.5 target (and if necessary document the 8.4 with a warning that "for now, entering the prefetch mode prevents using getColumnMeta() or anythng else meta calling the database's information schema").

This is a first, working pass on the subject, while I found time to code it: it is a technically **self-sustainable fix**, although we could go further in what `pgsql_finish_running_stmt()` is a first try at: separating functional PQexecs (the ones from `exec()`, passing the user queries) from technical ones (called internally by `getColumnMeta()` and such).
Then:
- any internal query, depending on a `mandatory` parameter, would either throw ("Cannot access the meta schema in the middle of a result consumption") or silently return nothing (e.g. in `getColumnMeta()`, if we cannot resolve a user-defined type to its pretty name, we can still return its OID).
- later, when implementing [libpq's pipeline mode](https://www.postgresql.org/docs/current/libpq-pipeline-mode.html), we could implement in this centralized function the complexity of taking advantage of the pipeline mode to direct those synchronous meta-select queries to a dedicated "channel" where they could execute even while the user-pulled result is in the middle of its fetch.

But as I am not sure of the time I could devote to this ideal solution,
**let's keep this PR focused on its first goal** of just preventing unexplainable libpq error codes.

outtersg commented:
@devnexen, @SakiTakamachi, I didn't see the time go by, and it seems to me that including this robustness PR would be a good thing to avoid dragging ourselves around with a feature still holding the risk of unexpectedly truncated results or even garbled connection: **the safeguards added in this PR** (before each `PQexec()`) **would be more reliable than simply documenting** that "some functions that may internally call the DB server should not be used while in `ATTR_PREFETCH` mode".

May you take time to reevaluate its cleanliness?

devnexen commented:
> @devnexen, @SakiTakamachi, I didn't see the time go by, and it seems to me that including this robustness PR would be a good thing to avoid dragging ourselves around with a feature still holding the risk of unexpectedly truncated results or even garbled connection: **the safeguards added in this PR** (before each `PQexec()`) **would be more reliable than simply documenting** that "some functions that may internally call the DB server should not be used while in `ATTR_PREFETCH` mode".
> 
> May you take time to reevaluate its cleanliness?

Hi @outtersg now is more timely, I ll a deeper look sometime this weekend, thanks for your effort (and patience).

devnexen commented on ext/pdo_pgsql/pgsql_driver.c:
```diff
@@ -109,6 +110,37 @@ int _pdo_pgsql_error(pdo_dbh_t *dbh, pdo_stmt_t *stmt, int errcode, const char *
+		 * and consume anyway (results consumption is handle-wise, so we have no formal
+		 * need for the statement). */
+		if (first) {
+			php_error_docref("ref.pgsql", E_NOTICE, "Internal error: unable to link a libpq result to consume, to its origin statement");
```
I think it should be `ref.pdo-pgsql` here

outtersg replied:
At the time, wondering what to put as a first param to `php_error_docref()`, probably I dumbly `grep -r php_error_docref(" ext/`ed (note the `"`) and copied the most approaching found label.

But in fact, while now `grep`ing _without_ the `"`, I see that _all_ occurrences in `ext/` use `NULL` as a first parameter, so [here we go](https://github.com/php/php-src/compare/d30395fa002cd73d5474e7020ee7e9bfd4096b92..be5c93ec11e2107e6749ebad2ff67fe0aa3605d1).

I still didn't RTFM, but I'll quickly suppose that:
* historically the parameter was required (we still find traces of such uses in `ext/iconv`, `ext/mbstring`, `ext/pgsql`, `ext/sqlite3`…)
* but the function then got able to fill a `NULL`-provided module (thus the more modern PDO exclusively relying on this autofill)
* the parameter is still useful when the error is not directly related to the extension (e.g. `ext/intl` using `NULL` for errors due to the inner workings of intl, but using `"session.configuration"` to tell about a configuration problem), or when the extension wants finer grained subcategories (e.g. `"session.gc_probability"`, `"session.gc_divisor"`, `"session.configuration"`)

devnexen replied:
yes good point

devnexen commented:
looking good minus the minor nit.

devnexen commented:
@outtersg unfortunately the fix brings memory issues, one memory leak and above all a UAF. I ll have to revert it for the time being so we have time to have a closer look.

outtersg commented:
> @outtersg unfortunately the fix brings memory issues, one memory leak and above all a UAF. I ll have to revert it for the time being so we have time to have a closer look.

OK, the memory leak is easy.
_Inside_ the function I [added an `estrdup` to the cache](https://github.com/php/php-src/pull/16249/files#diff-548f2e84f481651b4d7bfecbce74b1ada3afa6272a78438dba2f44aef6b56152R745) (in addition to the existing `estrdup` to the return value): 2 allocations.
But for the returned value, when _calling_ the function, when I replaced the use of the function's return value in `add_assoc_string()` by the use of the cached value, I removed the `efree` of the returned value. Thus we loose this `efree`, and only have 1 cleanup left (the one of the cache that has its own cycle, freing when a new cache takes its place, or on destruction). 2 allocations vs 1 free make the leak.

I think my intention was to only rely on `cached_table_name` and its long lifecycle. However the [direct use of it](https://github.com/php/php-src/pull/16249/files#diff-548f2e84f481651b4d7bfecbce74b1ada3afa6272a78438dba2f44aef6b56152R776) is **bad** design because `cached_table_name` should be considered private to `pdo_pgsql_translate_oid_to_table` (and to cleanup functions of course): a caller should never access it directly (supposing it has been updated in `pdo_pgsql_translate_oid_to_table()`), the only existing contract is that `pdo_pgsql_translate_oid_to_table()` returns the table name.

So in any case, the caller should use the returned value, not `cached_table_name`. What gets interesting is the relation between both:
* **as they are coded now, they are independent** (each one is allocated, and is (or should be) freed)\
  this is robust and respects principle of least astonishment (each caller that gets a frees it, period).\
  _However we get 3 allocations_ (one for `cached_table_name`, one for the returned `table_name`, and one internal to `add_assoc_string()`).
* however here the returned `table_name` is immediately consumed by `add_assoc_string()`, then freed. _If we can guarantee this evanescent use will always work this way_, we could simplify by telling that _`pdo_pgsql_translate_oid_to_table`'s returned string should always be copied elsewhere before the next call to `pdo_pgsql_translate_oid_to_table()`, or risk being overwritten.
